core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -89,21 +97,66 @@ private void addOrUpdateActivity(ActivityPayload payload, UserAnonymizedDTO user
 		else if (payload.endTime.getTime() - activity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
 				.getUpdateSkipWindow() || payload.startTime.before(activity.getStartTime()))
 		{
-			updateLastActivity(payload, userAnonymized, matchingGoal, activity);
-			cacheService.updateLatestActivityForUser(activity);
+			updateLastActivity(dayActivity, payload, userAnonymized, matchingGoal, activity);
+			cacheService.updateDayActivityForUser(dayActivity);
 		}
 	}
 
-	private Activity createNewActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
+	private ZonedDateTime getZonedStartOfDay(Date startTime, UserAnonymizedDTO userAnonymized)
 	{
-		return Activity.createInstance(userAnonymized.getID(), matchingGoal, payload.startTime, payload.endTime);
+		return startTime.toInstant().atZone(ZoneId.of(userAnonymized.getTimeZoneId())).truncatedTo(ChronoUnit.DAYS);
 	}
 
-	private void updateLastActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal,
-			Activity activity)
+	private ZonedDateTime getZonedStartOfWeek(Date startTime, UserAnonymizedDTO userAnonymized)
 	{
-		assert userAnonymized.getID().equals(activity.getUserAnonymizedID());
-		assert matchingGoal.getID().equals(activity.getGoalID());
+		ZonedDateTime zonedStartOfDay = getZonedStartOfDay(startTime, userAnonymized);
+		switch (zonedStartOfDay.getDayOfWeek())
+		{
+			case SUNDAY:
+				// take as the first day of week
+				return zonedStartOfDay;
+			default:
+				// MONDAY=1, etc.
+				return zonedStartOfDay.minusDays(zonedStartOfDay.getDayOfWeek().getValue());
+		}
+	}
+
+	private DayActivity createNewActivity(DayActivity dayActivity, ActivityPayload payload, UserAnonymizedDTO userAnonymized,
+			Goal matchingGoal)
+	{
+		if (dayActivity == null)
+		{
+			dayActivity = createNewDayActivity(payload, userAnonymized, matchingGoal);
+		}
+
+		Activity activity = Activity.createInstance(payload.startTime, payload.endTime);
+		dayActivity.addActivity(activity);
+		return dayActivity;
+	}
+
+	private DayActivity createNewDayActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
```

Comment:
If two threads concurrently conclude that the day activity is missing, we end up with two entities for one day. We need to figure out how to handle that. I propose to create a separate ticket for this.